test/e2e/common/node/pod_admission.go

```diff
@@ -65,6 +67,89 @@ var _ = SIGDescribe("PodOSRejection", framework.WithNodeConformance(), func() {
 	})
 })
 
+var _ = SIGDescribe("PodRejectionStatus", func() {
+	f := framework.NewDefaultFramework("pod-rejection-status")
+	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+	ginkgo.Context("Kubelet", func() {
+		ginkgo.It("should reject pod when the node didn't have enough resource", func(ctx context.Context) {
+			node, err := e2enode.GetRandomReadySchedulableNode(ctx, f.ClientSet)
+			framework.ExpectNoError(err, "Failed to get a ready schedulable node")
+
+			// Create a pod that requests more CPU than the node has
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "pod-out-of-cpu",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "pod-out-of-cpu",
+							Image: imageutils.GetPauseImageName(),
+							Resources: v1.ResourceRequirements{
+								Requests: v1.ResourceList{
+									v1.ResourceCPU: resource.MustParse("1000000000000"), // requests more CPU than any node has
+								},
+							},
+						},
+					},
+				},
+			}
+
+			pod = e2epod.NewPodClient(f).Create(ctx, pod)
+
+			// Wait for the scheduler to update the pod status
+			err = e2epod.WaitForPodNameUnschedulableInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace)
+			framework.ExpectNoError(err)
+
+			// Fetch the pod to get the latest status which should be last one observed by the scheduler
+			// before it rejected the pod
+			pod, err = f.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+			// force assign the Pod to a node in order to get rejection status later
+			binding := &v1.Binding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      pod.Name,
+					Namespace: pod.Namespace,
+					UID:       pod.UID,
+				},
+				Target: v1.ObjectReference{
+					Kind: "Node",
+					Name: node.Name,
+				},
+			}
+			err = f.ClientSet.CoreV1().Pods(pod.Namespace).Bind(ctx, binding, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			// kubelet has rejected the pod
+			err = e2epod.WaitForPodFailedReason(ctx, f.ClientSet, pod, "OutOfcpu", f.Timeouts.PodStartShort)
+			framework.ExpectNoError(err)
+
+			// fetch the reject Pod and compare the status
+			gotPod, err := f.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+			gotStatus := gotPod.Status
+			gomega.Expect(gotStatus.Phase).To(gomega.Equal(v1.PodFailed), "Pod should be in Failed phase")
+			gomega.Expect(gotStatus.Reason).To(gomega.HavePrefix("OutOf"), "Pod should be rejected due to OutOf*")
+			gomega.Expect(gotStatus.Message).ToNot(gomega.BeEmpty(), "Pod should have a rejection message")
+			gomega.Expect(gotStatus.StartTime).ToNot(gomega.BeNil(), "Pod should have a StartTime")
+			gomega.Expect(gotStatus.QOSClass).To(gomega.Equal(pod.Status.QOSClass), "QOSClass should be preserved")
+
+			// other fields should be dropped
```

Comment:
Sorry, I think there is a misunderstanding. I am suggesting this check to make sure that if we will add new field calculated on pod creation (like totalAllocatedRequests), this test will fail and the person adding it will be forced to either explicitly say that it must be cleared by pod admission or add it to the pod admission so it will be preserved. I don't think there is a generic solution here, this is why I am askig for this test.

This will be the description:

```suggestion
			// This detects if there are any new fields in Status that were dropped by the pod rejection.
			// These new fields either should be kept by kubelet's admission or added explicitly in the list of fields that are having a different value or must be cleared.
```

Reply:
Also, I also suggesting to change the check itself. Let me try to explain. Let's say we are adding a new field called "TotalAllocationResources" that is initialized on pod creation. This field is similar to the QoS - we want to preserve it's value after the pod rejection.

The only way to catch it in future if this test will compare the pod status when not assigned to the node with the pod status after rejection. So please return the check you had before where you do a deep compare of statuses.

When that new field TotalAllocationResources will be added, the test will fail, and the fix will be to do what you are doing in this PR for QoS.

So again, the kubelet logic with assigning of QoS explicitly is good.
The test needs to be updated to do deep comparison of two pod statuses.

Reply:
@SergeyKanzhelev updated.